### DISCUSSION TITLE
Mvn command does not report problems

### DISCUSSION THREAD

**Discussion by lukas-arb** [pmd/pmd#3495] [General]
Hi. I am trying to configure PMD on new spring boot project with maven. This is my pom.xml:

```xml
<plugin>
<groupId>org.apache.maven.plugins</groupId>
<artifactId>maven-pmd-plugin</artifactId>
<version>3.14.0</version>
<configuration>
	<linkXRef>false</linkXRef>
	<rulesets>
		<ruleset>/rulesets/java/braces.xml</ruleset>
		<ruleset>/rulesets/java/naming.xml</ruleset>
		<ruleset>/rulesets/java/basic.xml</ruleset>
		<ruleset>/rulesets/java/empty.xml</ruleset>
		<ruleset>/rulesets/java/imports.xml</ruleset>
		<ruleset>/rulesets/java/unnecessary.xml</ruleset>
		<ruleset>/rulesets/java/unusedcode.xml</ruleset>
	</rulesets>
	<failOnViolation>true</failOnViolation>
	<printFailingErrors>true</printFailingErrors>

</configuration>
<executions>
	<execution>
		<goals>
			<goal>check</goal>
		</goals>
	</execution>
</executions>
</plugin>
```

I also have a dummy controller with unused local variable:

```java
@GetMapping
  public ResponseEntity<String> sayHelloWorld() {
    int i = 0;
    return new ResponseEntity<>("Hello world", HttpStatus.OK);
  }
```

But whenever I run mvn pmd:pmd I don't get any errors for usused variables. Only pmd 7 deprecation warnings such as:

```
[WARNING] Discontinue using Rule name category/java/codestyle.xml/ForLoopsMustUseBraces as it is scheduled for removal from PMD. PMD 7.0.0 will remove support for this Rule.
[WARNING] Use Rule name category/java/codestyle.xml/ForLoopsMustUseBraces instead of the deprecated Rule name ...\target\pmd\rulesets\braces.xml/ForLoopsMustUseBraces. PMD 7.0.0 will remove support for this deprecated Rule name usage.
```

Any ideas what could be wrong?

**Comment by adangel** [comment]
You'll need to run `mvn verify` in order to make your build fail. The [check goal](https://maven.apache.org/plugins/maven-pmd-plugin/check-mojo.html) is bound to the verify lifecycle. It's important to call maven and pmd in this way so that the previous phases are executed (like compile, test-compile), as otherwise your project won't be compiled, which results in false positives/negatives since then PMD's typeresolution won't work correctly.

`mvn pmd:pmd` only executes pmd (without compiling your project), but does not check the results - you can manually inspect the result by looking at the file `target/pmd.xml`. But since your project might not have been compiled, the results are not trustworthy.

By the way, you are using rulesets from PMD 5 - while they still work (you'll see the rule deprecation warnings though), they definitely won't work anymore with PMD 7. So, it might be a opportunity for you to [create your own ruleset](https://pmd.github.io/latest/pmd_userdocs_making_rulesets.html) and select the rules you want (instead of just enabling all rules....). Enabling all rules might lead to a situation where one rule contradicts another rule.

**Comment by lukas-arb** [comment]
Hmm strange. In docs it says "`pmd:check verifies that the PMD report is empty and fails the build if it is not. This goal is executed by default when pmd:pmd is executed.`" thats why I thought, that it will execute verify as well. Thanks.

Right now it is checking my files under `src/main/java`, but it skips test files under `src/test/java` . Any ideas?

**Comment by adangel** [comment]
> "pmd:check verifies that the PMD report is empty and fails the build if it is not. This goal is executed by default when pmd:pmd is executed." 

Yeah, I noticed this sentence a couple of days ago. It just doesn't make sense. It is the other way round: pmd:pmd is executed by default when pmd:check is executed - or in other words: pmd:check triggers pmd:pmd....
For the upcoming m-pmd-p 3.15.0 I've fixed this description.

For checking test sources, there is an extra parameter: [includeTests](https://maven.apache.org/plugins/maven-pmd-plugin/pmd-mojo.html#includeTests) - it's by default disabled. You'll need to enable it explicitly.

**Comment by lukas-arb** [comment]
Such an easy fix. Thanks! Did not see this one and only noticed that my configuration skipped tests after running `pmd.bat`
